### ISSUE TITLE
Add an new input type to backfill gzipped logs

### ISSUE THREAD

**Issue by lminaudier** [elastic/beats#637]
Hi,

Following [this discussion on filebeat forum](https://discuss.elastic.co/t/backfilling-old-logs-with-filebeat/38428/2) I would like to ask if it is possible to implement a solution to easily backfill old gzipped logs with filebeat.

The proposed solution mentioned in the topic is to add a new dedicated `input_type`.

It is also mentioned in the topic that when filebeat reaches the end of input on `stdin` it does not give you the hand back and waits for new lines which makes things hard to script to perform backfilling.

What are your thoughts on this ?

Thanks for your hard work.

**Comment by ruflin** [comment]
I would see the implementation as following:

* Prospector: A gzip prospector would be added. Harvesters would only be opened based on filenames (no inode etc.). It is assume, that the files are not renamed and don't have to be tracked. A file has a completion state and if it is once completed, it is never read again. This simplifies the implementation. If a .gz file with a new filename is found, a harvester is started.
* Harvester: A gzip harvester is added. It unzips and reads the full file only once. After finishing reading the file, the harvester stops and is never started again. The offset is only stored, if it is interrupted in the middle of reading. In a first implementation, this could even be removed for simplicity. 

This would be nice to have but I think it is not on the top of our priority list. A community contribution here would be more then welcome.

For your second issue about running filebeat only until first completion, lets refer to this issue: https://github.com/elastic/filebeat/issues/219

**Comment by lminaudier** [comment]
Thanks for the fast reply and the pointer to the issue.

I will try to look at your implementation proposal. I am still quite new to Golang and the project, so I can't promise anything :)

**Comment by ruflin** [comment]
@lminaudier Always here to help.

**Comment by mryanb** [comment]
This would be a great feature addition. Currently the splunk-forwarder does something similar and will index log rotated files that have been gziped automatically.

**Comment by Ragsboss** [comment]
+1. Is anyone working on this? If not I could possibly take it up..

**Comment by ruflin** [comment]
@Ragsboss I don't think anyone is working on that. Would be great to have a PR for that to discuss the details.

**Comment by cFire** [comment]
@Ragsboss @ruflin I'm delighted to see there's someone looking to pick this up. Is this happening? The reason I ask is because it may be possible for me to spend some time helping out with this in lieu of building another solution for gziped logs to use internally.

**Comment by Ragsboss** [comment]
@cFire please feel free to take this up. I've started looking at the code familiarizing myself with Go in general and FileBeat code. But I haven't started the real work yet. I'll try to help you in anyway you want, just let me know.
Few thoughts I had. From a pure functional viewpoint, defining a new `input_type` doesn't seem ideal as that would force users to author a new prospector in the config file. Instead I felt it may be better for the code to automatically deal with compressed files as long as they match the given filename patterns in the config file. The code can instantiate a different prospectorer (IIUC) based on the file extension/type. But from implementation viewpoint if this is turning out to be difficult, I think it's ok to burden/ask the users for some extra config...

**Comment by ruflin** [comment]
From an implementation point of view I think we should go the route of having a separate input type. Currently the way filebeat is designed is that a prospector and harvester type are tightly coupled, so a prospector only starts one type of harvesters. It is ok if gzip harvester reuses lots of code from the log harvester (which I think it will) but as log tailing and reading a file completely only once are form my perspective two quite different behaviours. The question which will also be raised is if the gzip files will change their name over time (means have to be tracked based on inode / device) or it is enough to just store the filename and a read / unread flag in the registry.

**Comment by ruflin** [comment]
Here is the PR related to the above discussions: https://github.com/elastic/beats/pull/2227

**Comment by willsheppard** [comment]
We'd like this feature too.

Our main use of filebeat would be to take a set of rotated, gzipped logs that represent the previous day's events, and send them to elasticsearch. No tailing or running as a service is required.

A batch mode would also be good, but any other workaround is acceptable.

**Comment by ruflin** [comment]
@willsheppard Thanks for the details. For the batch mode you will be interested in https://github.com/elastic/beats/pull/2456 (in progress)

**Comment by ruflin** [comment]
Now that https://github.com/elastic/beats/pull/2456 is merged this feature got even more interesting :-)

**Comment by collabccie7** [comment]
Hello,

Has there been any update regarding the support for gzip files ? Please let know. 

Thanks.

**Comment by cFire** [comment]
Idk about the others, but I've not gotten any time to work on this.

**Comment by ruflin** [comment]
No progress yet on this from my side.

**Comment by willsheppard** [comment]
This gzip input filter would be the killer feature for us. We're being forced to consider writing an Elasticsearch ingest script from scratch which writes to the Bulk API, because we need to operate on logs in-place (no space to unzip them), and we would be using batch-mode (#2456) to ingest yesterday's logs from our web clusters.

**Comment by maddazzaofdudley** [comment]
Has there been any movement on this as this too is the killer feature for me too

**Comment by RJFoley** [comment]
Throwing in my support for this feature.

**Comment by plumpNation** [comment]
Would be an awesome feature to have.

**Comment by ruflin** [comment]
There is this open PR here that still needs some work and also involves quite some discussions: https://github.com/elastic/beats/pull/3070

**Comment by jordansissel** [comment]
> Harvesters would only be opened based on filenames (no inode etc.). 

@ruflin I believe inodes may to be tracked because `logrotate` (assuming this is a target use case) renames files and reuses file names. Unless another tracking mechanism (when is 'hello.txt.1.gz' a "new file"  below, for example)

Example:

```
% ls -il /tmp/hello.txt*
103196 -rw-rw-r--. 1 jls jls 12 Jan 24 03:17 /tmp/hello.txt
103131 -rw-rw-r--. 1 jls jls 32 Jan 24 03:16 /tmp/hello.txt.1.gz

% cat test.conf
/tmp/hello.txt {
  rotate 5
  compress
}

% logrotate -s /tmp/example.logrotate -f test.conf

% ls -il /tmp/hello.txt*
103218 -rw-rw-r--. 1 jls jls 32 Jan 24 03:17 /tmp/hello.txt.1.gz
103131 -rw-rw-r--. 1 jls jls 32 Jan 24 03:16 /tmp/hello.txt.2.gz
```

^^ Above, 'hello.txt.2.gz' is the same file (inode) as previous 'hello.txt.1.gz'.

We can probably achieve this without tracking inodes (tracking modification time and only reading .gz files after they have been idle for some small time?), but I think the filename alone is not enough because file names are reused, right?

**Comment by ruflin** [comment]
I exactly hit this issues during the implementation. That is why the implementation in https://github.com/elastic/beats/pull/3070 is not fully consistent with the initial theory in this thread.

The main difference now to a "normal" file is that it is expected, that a gz never changes, and if it changes, the complete file is read from the beginning again.

**Comment by jpdaigle** [comment]
+1 vote for gzip input

**Comment by smsajid** [comment]
+1 for gzip support

**Comment by ruflin** [comment]
@jpdaigle @smsajid Could you share more details on how exactly you would use this feature?

**Comment by jpdaigle** [comment]
@ruflin In our case, log files from hundreds of servers are streamed to a central log processor, which then outputs gzip-ed logs in chunks of a few seconds on a host that's accessible to developers. This is where a filebeat would come into play: prospect for the appearance of new .gz files, grab them and process them through a logstash pipeline.

The above is pretty "single use-case specific" though, in more generic terms, how we would use a .gz input filter is for grabbing the output of one logging system and "gluing" it to the input of a logstash pipeline.

**Comment by christiangalsterer** [comment]
We would also be interested in this feature as we also have scenarios where zipped files are generated by an intermediate system and where we can't read the plain files directly. It would be great if this can be combined with #4373 so that we can specify how many days back and/or based on file name patterns we can read zipped files.

**Comment by smsajid** [comment]
@ruflin My use case requires back-filling of gzipped log files. We have requirement to keep original log files for a minimum of 6 months. Right now, the only option for us is to unzip the log files from the central location to a staging location and read the files from there using file beats. This has the disadvantage that we need additional storage and regular cleanup of unzipped files once they are processed

**Comment by christiangalsterer** [comment]
Very similar requirement also here which seems to be not so uncommon in regulated environments.

**Comment by ruflin** [comment]
@jpdaigle @smsajid @christiangalsterer Thanks for sharing. The use case that you mention where you "only" have gzip files and no other files could actually be covered by https://github.com/elastic/beats/pull/3070 Now that we have prospectors better abstracted out, it should also be possible to make it a separate prospector type.

@christiangalsterer Which issue did you want to refer above? Not sure if the one you have in is the right one.

**Comment by strootman** [comment]
![Good Luck](http://i.imgur.com/lq65QMM.png)

**Comment by eperpetuo** [comment]
+1 for gzip support

**Comment by ruflin** [comment]
@eperpetuo Could you also share your detailed use case? I want to keep collecting data on how people would use the feature to then validate a potential implementation.

**Comment by eperpetuo** [comment]
@ruflin  In my case, during the log rotation, files are automatically gzipped. That is, the current log file is compressed and renamed to _foo.log.gz_ and a new _foo.log_ file is created. New log events start to be written in this new _foo.log_ file.

Now, I have experienced some delay during high throughput events. Imagine filebeat is 2 seconds behind the last line in the log file. When log rotation occurs and the file is gzipped, filebeat is not able to continue reading the file.

Although new lines written to the just created _foo.log_ file are perfectly collected by filebeat since from the beggining, last few lines of the now gzipped file are never shipped to Elasticsearch and in this case there is loss of information.

We currently use Splunk and it also presents some delay. However, the splunk-forwarder is able to collect all events even during log rotation and no message loss occurs, which is the most important.

This situation is preventing to move the solution to the production environment because we can't afford to lose messages.

**Comment by praseodym** [comment]
@eperpetuo For now, a workaround could be to set up logrotate to only gzip the second time a file is rotated (i.e. you end up with `log`, `log.1`, `log.2.gz`, `log.3.gz`, etc.). This is how many Linux distros do rotation as well.

**Comment by eperpetuo** [comment]
@praseodym Thanks. We'll look into this workaround.:metal:

**Comment by ruflin** [comment]
@eperpetuo If you disable `close_renamed` and `close_removed` filebeat should continue reading the file as it will keep it open. This should give you your expected behaviour. Feel free to open a topic on discuss for it to discuss further.

**Comment by jhnoor** [comment]
+1 for gzip support, we have a ton of xml backlogs that are gunzipped that I would like to send to logstash. Unzipping everything would take up 30x more space.

**Comment by sly4** [comment]
+1 for gzip support.  I have a pile of alerts.json.gz files I would like to re-run.

**Comment by brennentsmith** [comment]
+1 as well. Our use case is that our CDN provider (Edgecast) gzip's logs before shipping to us - as a result we have no way of receiving them raw. We could have a process that takes the gzipped objects, and decompress them to a second directory that filebeat watches, but that's frankly wasteful when were dealing with terabytes of logs.

**Comment by ruflin** [comment]
@sly4 @jhnoor @brennentsmith The cases you mentioned should be easier to cover as there is no overlap between zip and unzipped files. Thanks for sharing the uses cases.

**Comment by djjoshuad** [comment]
+1 on behalf of the DFIR world.  ELK is/was my go-to for ingesting large dumps of log data in whatever crazy format a customer provides them in.  I have no use for "tailing" a log file; I only need to read in static files.  Those files are almost always gzipped, and rarely is decompressing them a viable option.  As an example, the case I'm working right now involves (among many other things) about 200G of compressed log evidence spanning the last 18 months.  Decompression is not a viable option.  I'd love to adopt the beats way of doing things, if it becomes possible to do so.

**Comment by yodog** [comment]
+1

**Comment by georgezoto** [comment]
+1 for "reading gzipped files should be supported out of the box"

**Comment by kevin0211** [comment]
+1

**Comment by arenard** [comment]
Any updates on this ?

**Comment by iahmad-khan** [comment]
+1 for gz files as input

**Comment by mishat-realityi** [comment]
Are there any updates on this?

**Comment by cando-1p** [comment]
If I understand the documentation for filebeat [here](https://www.elastic.co/guide/en/beats/filebeat/current/configuration-filebeat-options.html#_literal_type_literal), filebeat can take logs from stdin or a udp socket.
So for most of the cases talked about hear, you could setup the config file appropriately and do something in bash like:
```
for f in *.gz ; do
   zcat $f | filebeat
done
```

The important use case I see for reading gzip files is logs that are rotated before they are shipped in the cases of a prolonged network outage.
For example on my busy server I have log rotation that happens every minute and I have delaycompress enabled. If there is a network outage to where filebeat sends the logs for more than a minute, It could end up missing logs because they were rotated into a gzip file.
I understand the difficulties of tracking between the original file and the gzipped file and you don't want to read logs from a gzipped file that the logs have already been read from. So I think it would be great if filebeat could take care of doing the rotating. :-D
That way it could update the registry info with the file that it caused to be gzipped.
The rotation could have nice rich support for rotating with date formatted path names and cleanup.

**Comment by C-Duv** [comment]
`filebeat`'s `-once` flag (« _Run filebeat only once until all harvesters reach EOF_ ») makes it stops at the end of stdin.

I just used it successfully with the following `filebeat` configuration file to import multiple old gzipped log files to my logstash instance:
```
filebeat.inputs:
- type: stdin
  enabled: true
  ignore_older: 0
  tags: ["web_access_log"]
  fields:
    server: "foo-server.example.com"
    webserver: "apache"
  fields_under_root: true
  json.overwrite_keys: true

output.logstash:
  hosts: ["logstash.example.com:8088"]
```

And with the following command:
```
zcat "file.gz" | filebeat -once -e -c "filebeat-std.config.yml"
```

**Comment by jundl77** [comment]
+1 for gzip support

**Comment by austriae** [comment]
+1 for gzip support

**Comment by jorgeeb4** [comment]
+1 for gzip support

**Comment by jasperdj** [comment]
After three years of waiting, we continue to **+1 for gzip support**

**Comment by mielie1000** [comment]
+1 for gzip support

**Comment by burandobata** [comment]
+1 for gzip support

**Comment by kirillrst** [comment]
+1

**Comment by GaganJotSingh** [comment]
Is there any update/progress for this issue?

**Comment by arnoldasb** [comment]
+1

**Comment by widhalmt** [comment]
+1 from me for gzip support

**Comment by ivan-mezentsev** [comment]
+1 for gzip support

**Comment by martinfmeneses** [comment]
+1 for gzip support

**Comment by obfuscatedc0de** [comment]
+1 for gzip support

**Comment by fenchu** [comment]
Gentlemen Splunk can read *.gz as is, and has for years. My problem is that my *.gz files are ~20GB and beyond my control. Currently I run in scheduler:
`7z.exe e -oc *.gz* | logstash.bat -f sensioproxy.grok`
or:
`7z.exe e -oc *.gz* |  filebeat.exe -once -e -c "sensioproxy-filebeat.yml"` to a logstash on elk-server

This works somewhat, but the `zcat|7z.exe e -oc` is way faster than the logstash|filebeat part so it takes all 16GB of memory on the clients before data has been transferred.

Another issue is handling of duplicates, since sincedb is not working. I currently have an ugly pythonscript to handle duplicates.

The logstash-codec-gzip_lines plugin does not output a proper format of my files like `zcat/7z.exe e -oc`

**Comment by GiovanniBattista** [comment]
+1 for gzip support

**Comment by caribbeantiger** [comment]
+1 for gzip support

**Comment by Umaigenomu** [comment]
+1 for gzip support

**Comment by jesusslim** [comment]
+1 for gzip support

**Comment by muthunagarajan** [comment]
+1 for gzip support

**Comment by sadokmtir** [comment]
+1 for gzip support

**Comment by holgerbrandl** [comment]
Log compression is the standard. It would be really great if filebeat could read gz-logs.

**Comment by kolikons** [comment]
+1 gzip

**Comment by esparky** [comment]
+1

**Comment by codingogre** [comment]
+1

**Comment by stevebanik-ndsc** [comment]
This issue has been open for 3 years.  Is the best approach to just use logstash for .gz files?

**Comment by incredincomp** [comment]
+1 gzip

**Comment by fenchu** [comment]
> This issue has been open for 3 years. Is the best approach to just use logstash for .gz files?
I recommend unpacking it before giving it to logstash. Another option is to use the _bulk endpoint to insert directly into elasticsearch, one line index and one line data at a far faster rate.

**Comment by ioigoume** [comment]
+1 gzip. 

Also a comment from my side is that consistency breaks if you create a new input type for zipped logs. I do understand that plain log files are different from gz files but it is much more "readable" to have one entry for all log files of the same application.

**Comment by mahmouddar** [comment]
any update :(

**Comment by IHeilig** [comment]
+1 for gzip

**Comment by amrithamenon16** [comment]
Is there any update on this?

**Comment by thekofimensah** [comment]
I have many log files that are in .gz format that I'd love to decompress automatically rather than running custom decompressing scripts first. I have 100's of file locations that I need to create scripts for that I wish I didn't have to.

**Comment by notestyle** [comment]
any update?

**Comment by notestyle** [comment]
+1 for .gz file

**Comment by alesanmed** [comment]
I think there are no news regarding this, right?

**Comment by lnmohankumar** [comment]
I wish this was available now. It can automate a lot of stuff.

**Comment by lnmohankumar** [comment]
I hope someone does progress on this.

**Comment by bunjamins** [comment]
agreed, looking for this functionality if possible.

**Comment by zxt47** [comment]
+1 please

**Comment by jbpratt78** [comment]
This would be an awesome feature :+1:

**Comment by Mobil92** [comment]
+1 for gzip please

**Comment by yodog** [comment]
what about https://github.com/elastic/beats/pull/2227 ?

**Comment by vprusa** [comment]
+1 for gzip please

**Comment by namishelex01** [comment]
+1 for gzip file processing

**Comment by lukeatt** [comment]
+1 for gzip please

**Comment by tjanson** [comment]
Please refrain from "+1" comments, it’s spam. Use the emoji reaction button if you want, but either way it won’t make development any faster.

**Comment by ITSEC-Hescalona** [comment]
+1 for gzip please

**Comment by notzippy** [comment]
As an alternative, you can use `mkfifo` and pipe that stream into filebeat and pipe all your files into that pipe.  That way you do not need to worry when you are finished ingesting a file because you can just automatically start piping the next file into it. So my `zcat`s go into `zcat ... >fb input` and in another terminal, I have `filebeat < fb_input`

**Comment by bunnHHA** [comment]
any update on gzip files? Using zcat solve some issues but it does need to be done in manual all the time.

**Comment by dwdixon** [comment]
+1 to adding this feature...is there any update on the timeline for this being added to Filebeat?

**Comment by jschin** [comment]
any update. +1 for gzip log input

**Comment by sachin-frayne** [comment]
+1

**Comment by NemchinovSergey** [comment]
+1

**Comment by arasic** [comment]
Hello everyone, 

The feature has not been implemented even if it has been requested by many users.
The ticket https://github.com/elastic/beats/pull/2227 was closed because there was no progress for some time. Some progess has been made, but I don't know whats remaining.

Not sure to understand if the feature was not implemented, is it because its a challenging problem or because lack of interest. 

From what I understood, there are two difficulties: 
* difficult to track/keep the offest of what has already been processed
* the files can be larger when extracted, like 50GBs and more?
* Did I miss any other challenging question?

So now, we realized that not only we cannot process the .gz files, but they should be excluded in exclude_files: [".gz"], otherwise we will get scrambled text.

The most helpful thing I found was posted by Vasek Šulc:
https://discuss.elastic.co/t/filebeat-on-demand-loading-compressed-gz-files-from-stdin/184159, which is to use the stdin configurations in filebeat, and redirect or pipe the compressed  file outputs using zcat.

**Comment by jlind23** [comment]
@nimarezainia Have you heard of this priority from any users so far?

**Comment by nerophon** [comment]
Competitors have this feature, e.g. Splunk forwarders, and also Graphana Loki: https://github.com/grafana/loki/issues/5956.

**Comment by nimarezainia** [comment]
this is tracked i the enhancements repo how ever not slated for development due to other higher priority items on the list.

**Comment by nimarezainia** [comment]
> Hi, is there any updates? Thanks

@fzyzcjy right now we don't have a development date for this.

**Comment by mathemaphysics** [comment]
This would be fabulous to have this feature. Accessing files directly would allow reading of log files directly without decompression of the file hierarchy. Text compression is generally very good, so this would be in line with preservation of disk space.

**Comment by matschaffer-roblox** [comment]
@jlind23 @nimarezainia incase it's helpful here's how I found this issue today:

We use a lot of AWS EMR where Amazon handles ingest of some logs (usually enough) but not all. Last weekend we hit a problem with one of the services in the mix that weren't ingested. We started a new EMR, scp'd most of the logs (some raw, some gz) from the old deployment and terminated the old one. Now I'd like to get the logs ingested for analysis.

ML's Data Visualizer does a great job on the raw files and even spits out a filebeat config for me. It'd be impressive if filebeat just worked for the .gz'd logs as well so I could slurp up everything I scp'd.

Tangentially having a filebeat option for https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html would also make this task easier.

Being able to quickly ingest and analyze a handful of local logs could be a nice on-ramp to continual ingestion as well, provided all the artifacts (filebeat config, ingest pipeline) were easily portable to something like a cloud deployment.

**Comment by philhagen** [comment]
This would be an immensely helpful feature.  👍

**Comment by pierrehilbert** [comment]
Done here: https://github.com/elastic/beats/issues/44866
